### ISSUE TITLE
Proposal to discuss: "Suggestion mode" cops

### ISSUE THREAD

**Issue by zverok** [rubocop-hq/rubocop#9069]
Thinking of [this Twitter discussion](https://twitter.com/tom_dalling/status/1329365392905678849) I have focused on a feeling I frequently have with Rubocop's suggestions, especially of `Style/` department. The thing is, the very same cop (say [Style/ReturnNil](https://docs.rubocop.org/rubocop/cops_style.html#stylereturnnil)) can be in one piece of code perceived a "oh, yeah, that's legacy/novice code/me being dumb", and in another one as "no, that's what I actually meant". The same might be related to lot of them, like `Style/IfInsideElse`, or some that are so agressive that they are disabled by default, like `Style/AutoResourceCleanup`.

The suggestions _might_ be useful (and sometimes I am geniunely entertained by "Oh, definitely, I could do it this way"), but also _might_ be annoying (and disabling every annoying place adds to visual noise, and to time for dealing with a new code).

The idea I have about them: Once in a while, I'd like to run rubocop on some codebase with an additional commandline option, like `rubocop --suggest`, and it will take those "controversial" cops and show what they suggest. (Ideally, there also should be a way to say "I've seen this suggestion and consider it irrelevant", like a cache of already shown... But probably it is too complicated, and can be solved by external tools like "run `--suggest` only on files changed from `git` perspective").

The flow is:
* You configure mandatory cops as `Enabled: true/false` and add Rubocop to CI
* You configure other cops as `Enabled: Suggest`, and they behave as disabled on `rubocop` and `rubocop -a`, and they (and only they) run on `rubocop --suggest` (and maybe they format themselves a bit differently, appropriate for "just suggesting, not being offended")

WDYT?..

**Comment by bbatsov** [comment]
That's an interesting proposal, but I've been thinking of tackling the problem a bit differently. Down the road I wanted to add another (configurable) level of grouping cops based on the importance/popularity of whatever they are about. There can be some conservative preset with fewer cops, something default with more, and something like "all in" (which is what we currently have).  It could be something like:

```
rubocop --preset essential
rubocop --preset style-guide
rubocop --preset github
```

I was thinking that this was we could also incorporate support for my style guides without the need for separate gems, but I haven't thought much how this would look from a configuration standpoint. Most like just different `.rubocop.yml`s bundled with RuboCop. 

We already have the notion of severity even today, but for some reason we never made much use of it. It could be another way to tackle this problem. I guess your idea is pretty similar when I think about it, as still someone has to go and decide on the importance of each cop.

**Comment by zverok** [comment]
Putting cops into groups seems more generic, yes (in fact, my feature request can be implemented with just `group: suggest` or something like that) -- **if** it would be easy to group/regroup them in local configuration file.

**Comment by marcandre** [comment]
@bbatsov are these levels like folders, or are they tags? They might be better as tags as some categories are orthogonal (e.g. beginner vs expert). But maybe that's too much freedom?

I've started a few small projects recently and I'm looking at what my favorite style is, and I don't think that these presets would help that much (so far). Most of the times, it was choices that are just personal and I simply disagree with RuboCop's default (e.g. trailing comma!!!).

It's not clear to me that curated lists of settings aren't what is needed instead (and `standard` can be seen as one). Maybe all we need is a list of "name" => "URL" for those?

Alternatively I'm wondering if a list of the most "personal-choice" cops wouldn't be more helpful, like an interractive questionnaire of the top 10 questions?

>  Style/ReturnNil

That cop is disabled by default, and I'm glad because that's a really bad cop imo. What preset did you think it should belong to?

> Style/IfInsideElse

Curious example; I can't think of anybody having a good reason to disable this cop. The choice between a bunch of `elsif` and a `case` (again, I disagree with RuboCop's choice here) is difficult, but `IfInsideElse` seems to me like it should always be enforceable.

**Comment by zverok** [comment]
>> `Style/IfInsideElse`
>
> Curious example; I can't think of anybody having a good reason to disable this cop. The choice between a bunch of elsif and a case (again, I disagree with RuboCop's choice here) is difficult, but IfInsideElse seems to me like it should always be enforceable.

I typically not disable it, but sometimes I feel uneasy about its suggestion. Can't remember the exact project/situation (and actually I prefer `case` too), but in some complicated algo, you have:
```ruby
if some_domain.one_condition?
  # ...
elsif the_same_domain.another_conditioin?
  # ...
elsif still_the_same_domain.third_condition?
  # ...
else
  # fallback for outer domain, but NOW we have some additional stuff to check
  if nested_different_domain.one_condition?
  elsif that_nested_domain.another_condition?
  else
  end
end
```
And yeah, "you should split it in methods" and such stuff, but in the vacuum, talking about only that cop, the code above might be considered clearer than
```ruby
if some_domain.one_condition?
  # ...
elsif the_same_domain.another_conditioin?
  # ...
elsif still_the_same_domain.third_condition?
  # ...
elsif nested_different_domain.one_condition? # "now we start checks in different domain" is lost here
  # ...
elsif that_nested_domain.another_condition?
  # ...
else
  # ...
end
```

**Comment by bbatsov** [comment]
> @bbatsov are these levels like folders, or are they tags? They might be better as tags as some categories are orthogonal (e.g. beginner vs expert). But maybe that's too much freedom?

I was thinking of tags, but I didn't think that much about this yet. It's just something that's been on the back on my mind for a while now, as from to time to time someone has been complaining that the default config has controversial cops in it and wanted a more conservative config. I also see a ton of `rubocop-company-name` gems that simply exist to host one configuration file. Probably we can do better on both fronts.

**Comment by bbatsov** [comment]
> That cop is disabled by default, and I'm glad because that's a really bad cop imo. What preset did you think it should belong to?

No idea. That was an external contribution if I recall correctly, and it's quite controversial therefore disabled. I guess someone finds it useful if they decided to work on it. I don't think that every cop should be in some preset, as some cops are kind of mutually exclusive.

**Comment by marcandre** [comment]
> > That cop is disabled by default, and I'm glad because that's a really bad cop imo. What preset did you think it should belong to?
> 
> No idea. That was an external contribution if I recall correctly, and it's quite controversial therefore disabled. I guess someone finds it useful if they decided to work on it. I don't think that every cop should be in some preset, as some cops are kind of mutually exclusive.

I should have tagged @zverok since he brought it up as an example...

**Comment by zverok** [comment]
>>> That cop is disabled by default, and I'm glad because that's a really bad cop imo. What preset did you think it should belong to?
>>
>> No idea. That was an external contribution if I recall correctly, and it's quite controversial therefore disabled. I guess someone finds it useful if they decided to work on it. I don't think that every cop should be in some preset, as some cops are kind of mutually exclusive.
>
> I should have tagged @zverok since he brought it up as an example...

Actually, it is was @tomdalling's example in the blog post :)

But, TBH, in some codebases I like to have it enabled: while I understand the concerns about explicitness, the "implicit returned `nil`" gives me a warm feeling of "minimal way to express the meaning", which I really love about Ruby. Yeah, `return` without value is somewhat obscure idiom, but still.

**Comment by marcandre** [comment]
FWIW, `rbs` distinguishes between a return type of `nil`, or `SomeClass?` and a `void` return type (that can be technically be any value, but result is not meant to be used). I really like this `void` signature idea; I `return nil` when the `nil` should be used, and I `return` when the method's result is not used (`void`)

Anyways, I think a list of cops that are more typically a user preference would be good; some are also separate but really are the same issue in my mind (trailing commas!!!).

**Comment by tomdalling** [comment]
Either using presets or severity/suggestions could be good. But you still have the same problem: people are going to interpret the defaults as the "correct" way to write Ruby. So what would the defaults be?

Personally, I think the defaults should be more relaxed and unobtrusive. I feel like, in Ruby, strictness should be optional for most things --- not just for RuboCop. Ruby is supposed to be flexible compared to other languages, and strict config takes that away. And if you're trying to avoid the complainy people like myself, relaxing the defaults is the way to go. If it's not a default then you can't blame RuboCop, you can only blame the person who configured it to be that way.

But it's not about what I personally would like. I think you just need to decide what the spirit (for lack of a better word) of RuboCop is. Does RuboCop want to be strict by default, to encourage consistency, letting people relax the config if they want to? Or does RuboCop want to be relaxed by default, to be unobtrusive and allow flexibility, letting people tighten the config if they want to? There's no wrong answer, but I don't think the rules can be both unobtrusive and highly consistent at the same time.

> That cop is disabled by default, and I'm glad because that's a really bad cop imo. What preset did you think it should belong to?

Possibly something like a "mutation testing" preset. Mutation testing tends to remove everything that can possibly be removed, so `Style/ReturnNil` and `Style/EmptyElse` would be preferable there.

**Comment by bbatsov** [comment]
I'm leaning towards a more relaxed default preset myself. 

> And if you're trying to avoid the complainy people like myself, relaxing the defaults is the way to go. If it's not a default then you can't blame RuboCop, you can only blame the person who configured it to be that way.

Largely because of this. 😆

**Comment by bbatsov** [comment]
Closing as stale — interesting discussion about 'suggestion mode' cops but no consensus was reached and it has been dormant for 5+ years. If someone wants to revive this, please open a fresh proposal.
